### ISSUE TITLE
Angular Elements: `document-register-element` breaking change 

### ISSUE THREAD

**Issue by stevermeister** [angular/angular#24390]
## I'm submitting a...

<pre><code>
[x ] Bug report  
</code></pre>

## Current behavior
**Angular Elements**

This package has breaking change -
https://github.com/angular/angular/blob/master/packages/elements/schematics/ng-add/index.ts#L34

After it's automatically updated to *1.9.0*. It gives an error in browser:

> Error: Failed to construct 'HTMLElement': Please use the 'new' operator, this DOM object constructor cannot be called as a function.

## Environment

<pre><code>
Angular version: 6.0.3

Browser:
- [x ] Chrome (desktop) version 67

</code></pre>

**Comment by JenPeleva** [comment]
This might help: https://github.com/stackblitz/core/issues/475
Or setting the 'target' in your tsconfig.json to es2015

**Comment by stevermeister** [comment]
@JenPeleva thank you for the hint! 
But I still think that current functionality should be hotfix in core
that's why I created this hotfix pull-request - https://github.com/angular/angular/pull/24391

**Comment by WebReflection** [comment]
alternatively, this is what landed in Babel 7:
https://github.com/WebReflection/babel-plugin-transform-builtin-classes

**Comment by KHappe** [comment]
So what is the fix for this issue?

**Comment by WebReflection** [comment]
@KHappe you need to transpire code properly, in a way that doesn't break native classes extends.

**Comment by KHappe** [comment]
@WebReflection I'm just using the default setting generated by @angular/cli.  Do I need to change tsconfig.json?  What change do I need to make to have it transpile code correctly?

**Comment by WebReflection** [comment]
@KHappe if this project uses TypeScript I've no idea how to solve it, so the most pragmatic solution for the time being is pinning the DRE version to 1.8.1 which is the most updated version of the library before introducing changes that would break latest Chrome if the code is badly transpiled.

That being said, I suggest this project to find a solution to properly transpile classes, filing bugs at TypeScript if necessary, saying that this fails:

```js
class AFailure extends HTMLElement {}
customElements.define('a-failure', AFailure);
const fail = new AFailure;
```

as [this link](http://www.typescriptlang.org/play/index.html#src=class%20AFailure%20extends%20HTMLElement%20%7B%7D%0D%0AcustomElements.define('a-failure'%2C%20AFailure)%3B%0D%0Aconst%20fail%20%3D%20new%20AFailure%3B) easily demonstrates (try that transpiled code in Chrome or Safari and see same failure)

The alternative is to not change classes in TypeScript but I don't know what are your target browsers.

**Comment by WebReflection** [comment]
More on this topic, since it's a well known TypeScript issue since about ever:
https://github.com/Microsoft/TypeScript/issues/15397

there are some possible solutions in here too:
https://github.com/Microsoft/TypeScript/issues/12949

**Comment by njl07** [comment]
Same problem for me and downgrade `document-register-element` to `v1.8.1` solved it

**Comment by KHappe** [comment]
V1.8.1 worked for me too.  Thanks for finding that!!!

**Comment by tndk1995** [comment]
it worked for me too! Thanks :D

**Comment by njl07** [comment]
You can try my library to load custom elements if you want https://github.com/hapinessjs/ng-elements-loader

**Comment by PeterKassenaar** [comment]
Pinning DRE to 1.8.1 does not work for me, not even with a blank new CLI project out of the box.

No element can be registered, I always get the error "Uncaught TypeError: Failed to construct 'HTMLElement': Please use the 'new' operator, this DOM object constructor cannot be called as a function." 

So Angular Elements is basically unusable at this point?

**Comment by WebReflection** [comment]
I'd like to underline this is not directly an issue of DRE. DRE was using a feature detection on Chrome that was failing, because Chrome supports native classes, native Custom Elements V0, and native Custom Elements V1, including builtins extend.

As soon as Chrome fixes that funny feature detection failure on checking the `is` attribute, the DRE, based on feature detections, will stop patching Chrome, and this is expected, but also, correct, from a polyfill perspective.

What you need to do is to switch to Babel 7 or transpile at least your output **before** TypeScript, through Babel 6, and this plugin: https://github.com/WebReflection/babel-plugin-transform-builtin-classes (which is the one basically merged into Babel 7).

**Comment by WebReflection** [comment]
Alternatively, you need to target es2015 instead of es5, so that classes are left untouched the [support is 88.31%](https://caniuse.com/#feat=es6-class), and you'll live happily ever after with classes from that time on.

**Comment by WebReflection** [comment]
**another alternative** is to force patch customElements no matter what:
https://github.com/WebReflection/document-register-element#avoiding-ce-built-in

```js
const installCE = require('document-register-element/pony');

// by default, the second argument is 'auto'
// but it could be also 'force'
// which ignores feature detection and force
// the polyfill version of CustomElements
installCE(global, 'force');
```

please note I am using the `/pony` fill, not the polyfill.

**Comment by PeterKassenaar** [comment]
@WebReflection Thanks for the clarification - but I tried to set the options in my main tsconfig.json to "es2015" and still no go. Weird. 

I think I understand what is going wrong and why, but I can't get my (basic) Angular project to compile correctly. It's using TS 2.7.2 by default.

**Comment by WebReflection** [comment]
@PeterKassenaar then try:

```js
import {dre} from 'document-register-element/pony';
dre(window, 'force');
```

if that works, it is **not** a long term solution, but at least it should enable this project.

**Comment by jessndaa** [comment]
Hi, I have solved this issue by changing the "target":"es5" in the tsconfig.json to "target":"es2015"
these is my tsconfig.json file
```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es2015",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ]
  }
}```

**Comment by PeterKassenaar** [comment]
@jessndaa @WebReflection I updated the tsconfig.json to 'es2015' and somehow it did *not* work in the project when compiling and building the project. However, when I copied over the /src folder (and package.json) to a brand new project, it  *did* work.

So I don't have a clue for the Why and How, but I'm happy for now.

Thanks again for your help and time.

**Comment by renebizelli** [comment]
V1.8.1 worked for me too. Tks!!!!

**Comment by haydenbr** [comment]
@jessndaa's solve worked for me. On angular 6.1.6, that's the only change I had to make and it worked fine.

**Comment by njl07** [comment]
If your application is built in `es5` you can just add a new polyfills `@webcomponents/webcomponentsjs/custom-elements-es5-adapter` and all will work without problem.

`document-register-element` is only required if the browser doesn't support `HTMLElement`

**Comment by WebReflection** [comment]
> document-register-element is only required if the browser doesn't support HTMLElement and can be in latest version now with the previous polyfill

No. document-regidster-element is the most deployed polyfill for Custom Elements out there, and it doesn't require any particular adapter, it works already with Babel 7 out of the box.

**Comment by WebReflection** [comment]
P.S. since people still write in here, feel free to read how to migrate to V1, requiring zero polyfills if not, eventually, the one for built-in elements.
https://github.com/WebReflection/document-register-element#important-how-to-migrate-to-native-v1

**Comment by WebReflection** [comment]
there is a fix for babel 6 too, it's the same fix that eventually landed in babel 7
https://github.com/WebReflection/babel-plugin-transform-builtin-classes#readme

> you don’t need it if your browser already support all features.

neither you need anything else ... if the browser natively supports customElements it's granted to support also native classes so that no poly is needed.

Again, this is explained in the link posted in 
https://github.com/angular/angular/issues/24390#issuecomment-437330016

**Comment by njl07** [comment]
> there is a fix for babel 6 too, it's the same fix that eventually landed in babel 7

I'm sorry but I don't use `Babel` but only `ts-loader` so I've the problem and must have a polyfill.

> neither you need anything else ... if the browser natively supports customElements it's granted to support also native classes so that no poly is needed.

It's not true because my browser is the latest version of `Chrome` and I've the problem when I build with `webpack` and it's not the problem of `document-register-element` but of the `es5` compilation that needs `custom-elements-es5-adapter`

**Comment by njl07** [comment]
@WebReflection I've just trying your solution [babel-plugin-transform-builtin-classes](https://github.com/WebReflection/babel-plugin-transform-builtin-classes#usage) and it doesn't work for me.

### webpack.config.js
```js
var path = require('path');

module.exports = {
    entry: { main: './main.ts' },
    resolve: { extensions: [ '.ts', '.js' ] },
    target: 'web',
    // this makes sure we include node_modules and other 3rd party libraries
    externals: [ /(node_modules)/ ],
    output: {
        path: path.join(__dirname),
        filename: '[name].js'
    },
    module: {
        rules: [
            {
                test: /\.tsx?$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            presets: [ 'babel-preset-env' ]
                        }
                    },
                    {
                        loader: 'ts-loader'
                    }
                ]
            }
        ]
    },
    stats: {
        warnings: false
    }
};
```

### .babelrc
```json
{
  "plugins": [
    "babel-plugin-transform-es2015-classes",
    ["babel-plugin-transform-builtin-classes", {
      "globals": ["Array", "Error", "HTMLElement"]
    }]
  ]
}
```

### package.json
```json
{
"devDependencies": {
    "babel-core": "^6.26.3",
    "babel-loader": "^7.1.5",
    "babel-plugin-transform-builtin-classes": "^0.6.1",
    "babel-plugin-transform-es2015-classes": "^6.24.1",
    "babel-preset-env": "^1.7.0",
    "ts-loader": "^5.3.0",
    "webpack": "^4.23.1",
    "webpack-cli": "^3.1.2"
  }
}
```

**Comment by njl07** [comment]
@WebReflection I've just trying with `Babel7` it doesn't work anymore.

### webpack.config.js
```js
var path = require('path');

module.exports = {
    entry: { main: './main.ts' },
    resolve: { extensions: [ '.ts', '.js' ] },
    target: 'web',
    // this makes sure we include node_modules and other 3rd party libraries
    externals: [ /(node_modules)/ ],
    output: {
        path: path.join(__dirname),
        filename: '[name].js'
    },
    module: {
        rules: [
            {
                test: /\.tsx?$/,
                use: [
                    {
                        loader: 'ts-loader'
                    },
                    {
                        loader: 'babel-loader',
                        options: {
                            presets: ['@babel/preset-env']
                        }
                    }
                ]
            }
        ]
    },
    stats: {
        warnings: false
    }
};
```

### package.json
```json
{
"devDependencies": {
    "@babel/core": "^7.1.5",
    "@babel/preset-env": "^7.1.5",
    "babel-loader": "^8.0.4",
    "ts-loader": "^5.3.0",
    "webpack": "^4.23.1",
    "webpack-cli": "^3.1.2"
  }
}
```

**Comment by njl07** [comment]
So I think the problem isn't with `document-register-element` but with `@angular/elements` itself because first one isn't include in my file and I'm just calling `angular` feature.

If @robwormald can check it'll be cool, thanks to him

**Comment by njl07** [comment]
If I'm in an `Angular` application and I'm setting `target: es2015` in my `tsconfig.json` file all works fine without any `polyfills`.

When I'm using `@angular/elements` outside `Angular` and built with `webpack` and `ts-loader`, I must have `@webcomponents/webcomponentsjs/custom-elements-es5-adapter`polyfill.

It'll be cool if we can have the `webpack` configuration used in `Angular-CLI` to work properly with `@angular/elements` outside `Angular`

**Comment by WebReflection** [comment]
> the es5 compilation

I think you should never do that ... you should [serve es5 only to legacy browsers](https://itnext.io/a-bloatless-web-d4f811c7991b), not all of them.

Also, about using typescript, if you transpire to ES5 classes are plain broken when it comes to builtin extends.

This is a [long standing bug](https://github.com/Microsoft/TypeScript/issues/15397) that won't get fixed. [Watch out your tools](https://itnext.io/js-is-not-worse-than-84a5fca0e87d#fd9a).

**Comment by ChrisTheButcher** [comment]
Fix for ES5 as a target:

Add this to your `polyfills.ts` on the bottom of your file
```
/***************************************************************************************************
 * APPLICATION IMPORTS
 */
import '@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js'; 
```

make sure to have the module installed with `npm install @webcomponents/webcomponentsjs --save-dev`

**Comment by renevall** [comment]
Fix does not work for me sadly, just the changing target.

**Comment by blackat** [comment]
[Solution ](https://github.com/stackblitz/core/issues/475) that worked for me, thanks to @robwormald 

Add this to your index.html VanillaJS page where you use the generated tag:
https://unpkg.com/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js

**Comment by CSchulz** [comment]
it seems using the `custom-elements-es5-adapter.js` makes the `document-register-element` script obsolete.

I am not familiar with both libraries, what are the main differences?

**Comment by ChristianUlbrich** [comment]
@CSchulz No it does not. `custom-elements-es5-adapter.js` simply allows **transpiled** ES5 code to utilize the **native** CustomElement API that for some unknown reason demands ES2015+ _class syntax_, whereas the `document-register-element` is polyfilling the CustomElement API.

**tl;dr;** You need `document-register-element` for legacy browsers not supporting it, i.e. IE11, Firefox 3, ... **and** `custom-elements-es5-adapter.js` for the ones supporting it, if you are using (_transpiled_) ES5 code, which you likely do.

**Comment by EugenMayer** [comment]
I am not entirely sure i contribute this way, but maybe this helps. Ran into the same issue, having target type es2015 and using 1.13.1 of DSR and was running into the mentioned issue.

It turned out i was not loading `scripts.js` which was added by DSL with `ng add` and i was not aware of. adding this fixed the issue right away.

**Comment by gkalpak** [comment]
AFAICT, this issue is now obsolete. The latest versions of the CLI configure TS to target ES2015 by default and uses babel 7 to downlevel to ES5 (if needed), as described in the [Defferential Loading][1] guide.

If someone wants to explicitly configure TS to target ES5, they should be responsible for ensuring their polyfills work. Possible options (mentioned above):
- Force the `document-register-element` po[ln]yfill (as described in https://github.com/angular/angular/issues/24390#issuecomment-398781359 or https://github.com/angular/angular/issues/24390#issuecomment-398782603).
- Use an adapter, such as [native-shim.js][2] from `@webcomponents/custom-elements`.
- Use a different polyfill that works with "badly" transpiled classes.

##
I am going to close the issue as obsolete, but feel free to comment below if I missed something.

[1]: https://v10.angular.io/guide/deployment#differential-loading
[2]: https://unpkg.com/browse/@webcomponents/custom-elements@1.4.3/src/native-shim.js

**Comment by WebReflection** [comment]
FYI the modern version of `document-register-element` is [@ungap/custom-elements](https://github.com/ungap/custom-elements#readme) which is lighter, faster, better, stronger ... it fixes `whenDefined` to latest standards, and it patches down to IE11, it fixes Safari missed builtin extends, and it leaves untouched Chrome, Firefox, and modern Edge 👋
